src/com/palmergames/bukkit/towny/command/NationCommand.java

```diff
@@ -188,43 +204,110 @@
 	@Override
 	public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
 
-		if (args.length == 1) {
-			List<String> nationNames = NameUtil.getNationNames();
-			nationNames.addAll(nationTabCompletes);
-			return NameUtil.filterByStart(nationNames, args[0]);
-		}
-
-		if (args.length == 2) {
-			switch (args[0].toLowerCase()) {
-				case "set":
-					return NameUtil.filterByStart(nationSetTabCompletes, args[1]);
-				case "rank":
-				case "ally":
-				case "enemy":
-					return NameUtil.filterByStart(new ArrayList<>(Arrays.asList(
-						"add",
-						"remove"
-					)), args[1]);
-				case "kick":
-				case "add":
-					return NameUtil.getTownNamesStartingWith(args[1]);
-				case "toggle":
-					return NameUtil.filterByStart(nationToggleTabCompletes, args[1]);
-				case "join":
-					return NameUtil.getNationNamesStartingWith(args[1]);
-
+		if (sender instanceof Player) {
+			Player player = (Player) sender;
+			switch (args.length) {
+				case 1:
+					// Only suggest nation names in tab complete if the player is not typing a valid sub command. This reduces clutter and the unnecessary checking of all nation names for tab completion
+					List<String> nationNames = NameUtil.filterByStart(nationTabCompletes, args[0]);
+					if (nationNames.size() > 0) {
+						return nationNames;
+					} else {
+						return NameUtil.getNationNamesStartingWith(args[0]);
+					}
+				case 2:
+					switch (args[0].toLowerCase()) {
+						case "set":
+							return NameUtil.filterByStart(nationSetTabCompletes, args[1]);
+						case "rank":

```

Comment:
The rank will have to return a list of players in the nation

Reply:
It does farther down, see the following [lines](https://github.com/TownyAdvanced/Towny/pull/3710/files/202c0e6584096a036b9cf6468cc4450f1812b726#diff-7e716e72bf8f5bffab172ac2a81e7a83R265-R276) around 270-ish. 

I'm not sure if I want to go with this style, but for the first `switch` statement I have cases that return possible sub commands. If none of the cases return anything, the second `switch` inside of the `if (args.length > 1) {` will handle returning possible residents, towns, nations, ranks, perms, etc. 

I may want to everything in one `switch` which has cases for each sub command. This may be easier in terms of readability with especially long commands, such as `/t set perm ally build off` as an example